src/api/tap.rs

```diff
@@ -0,0 +1,401 @@
+use crate::{
+    event::{Event, LogEvent},
+    topology::{fanout, WatchRx},
+};
+use futures::{future::try_join_all, FutureExt, Sink};
+use itertools::Itertools;
+use std::collections::VecDeque;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::{
+    collections::{HashMap, HashSet},
+    iter::FromIterator,
+};
+use tokio::sync::{mpsc as tokio_mpsc, mpsc::error::SendError, oneshot};
+use uuid::Uuid;
+
+/// A tap sender is the control channel used to surface tap payloads to a client.
+type TapSender = tokio_mpsc::Sender<TapPayload>;
+
+/// Clients can supply glob patterns to find matched topology components.
+trait GlobMatcher<T> {
+    fn matches_glob(&self, rhs: T) -> bool;
+}
+
+impl GlobMatcher<&str> for String {
+    fn matches_glob(&self, rhs: &str) -> bool {
+        match glob::Pattern::new(self) {
+            Ok(pattern) => pattern.matches(rhs),
+            _ => false,
+        }
+    }
+}
+
+/// A tap notification signals whether a pattern matches a component.
+#[derive(Debug)]
+pub enum TapNotification {
+    Matched,
+    NotMatched,
+}
+
+/// A tap payload can either contain a log/metric event or a notification that's intended
+/// to be communicated back to the client to alert them about the status of the tap request.
+#[derive(Debug)]
+pub enum TapPayload {
+    Log(String, LogEvent),
+    Metric(String, LogEvent),
+    Notification(String, TapNotification),
+}
+
+impl TapPayload {
+    pub fn matched(input_name: &str) -> Self {
+        Self::Notification(input_name.to_string(), TapNotification::Matched)
+    }
+
+    pub fn not_matched(input_name: &str) -> Self {
+        Self::Notification(input_name.to_string(), TapNotification::NotMatched)
+    }
+}
+
+/// A `TapSink` is used as an output channel for a topology component, and receives
+/// `Event`s. If these are of type `Event::LogEvent`, they are relayed to the tap client.
+pub struct TapSink {
+    tap_tx: TapSender,
+    component_name: String,
+    buffer: VecDeque<LogEvent>,
+}
+
+impl TapSink {
+    pub fn new(tap_tx: TapSender, component_name: String) -> Self {
+        Self {
+            tap_tx,
+            component_name,
+            // Pre-allocate space of 100 events, which matches the default `limit` typically
+            // provided to a tap subscription. If there's a higher log volume, this will block
+            // until the upstream event handler has processed the event. Generally, there should
+            // be little upstream pressure in the processing pipeline.
+            buffer: VecDeque::with_capacity(100),
+        }
+    }
+}
+
+impl Sink<Event> for TapSink {
+    type Error = ();
+
+    /// The sink is ready to accept events if buffer capacity hasn't been reached.
+    fn poll_ready(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if self.buffer.len() == self.buffer.capacity() {
+            Poll::Pending
+        } else {
+            Poll::Ready(Ok(()))
+        }
```

Comment:
In general this would be roughly correct, but we have a requirement for this sink to never block (i.e. you should never slow down normal processing because of an active `tap`). That means we have to always return `Ready` here.

Reply:
Moved to a non-blocking approach in ab6918383207cffea32b19c29223ab848317f601.